internal/externaldns/sync.go

```diff
@@ -0,0 +1,238 @@
+		}
+		return nil
+	}
+}
+
+func getValidTargets(endpoints []vsapi.ExternalEndpoint) (extdnsapi.Targets, string, error) {
+	var targets extdnsapi.Targets
+	var recordType string
+	var recordA bool
+	var recordCNAME bool
+	var recordAAAA bool
+	var err error
+	glog.Infof("Going through endpoints %v", endpoints)
+	for _, e := range endpoints {
+		if e.IP != "" {
+			glog.V(3).Infof("IP is defined: %v", e.IP)
+			if errMsg := validators.IsValidIP(e.IP); len(errMsg) > 0 {
+				continue
+			}
+			ip := netutils.ParseIPSloppy(e.IP)
+			if ip.To4() != nil {
+				recordA = true
+			} else {
+				recordAAAA = true
+			}
+			targets = append(targets, e.IP)
+		} else if e.Hostname != "" {
+			glog.V(3).Infof("Hostname is defined: %v", e.Hostname)
+			targets = append(targets, e.Hostname)
+			recordCNAME = true
+		}
+	}
+	if len(targets) == 0 {
+		return targets, recordType, errors.New("valid targets not defined")
+	}
+	if recordA {
+		recordType = recordTypeA
+	} else if recordAAAA {
+		recordType = recordTypeAAAA
+	} else if recordCNAME {
+		recordType = recordTypeCNAME
+	} else {
+		err = errors.New("recordType could not be determined")
+	}
+	return targets, recordType, err
+}
+
+func buildDNSEndpoint(extdnsLister extdnslisters.DNSEndpointLister, vs *vsapi.VirtualServer, targets extdnsapi.Targets, recordType string) (newDNSEndpoint, updateDNSEndpoint *extdnsapi.DNSEndpoint, _ error) {
+	existingDNSEndpoint, err := extdnsLister.DNSEndpoints(vs.Namespace).Get(vs.Spec.Host)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil, nil, err
+		}
+	}
+	var controllerGVK schema.GroupVersionKind = vsGVK
+
+	dnsEndpoint := &extdnsapi.DNSEndpoint{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            vs.Spec.Host,
+			Namespace:       vs.Namespace,
+			Labels:          vs.Labels,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(vs, controllerGVK)},
+		},
+		Spec: extdnsapi.DNSEndpointSpec{
+			Endpoints: []*extdnsapi.Endpoint{
+				{
+					DNSName:          vs.Spec.Host,
+					Targets:          targets,
+					RecordType:       buildRecordType(vs, recordType),
+					RecordTTL:        buildTTL(vs),
+					Labels:           buildLabels(vs),
+					ProviderSpecific: buildProviderSpecificProperties(vs),
+				},
+			},
+		},
+	}
+
+	vs = vs.DeepCopy()
+
+	if existingDNSEndpoint != nil {
+		glog.V(3).Infof("DNDEndpoint already exist for this object, ensuring it is up to date")
+		if metav1.GetControllerOf(existingDNSEndpoint) == nil {
+			glog.V(3).Infof("DNSEndpoint has no owner. refusing to update non-owned resource")
+			return nil, nil, nil
+		}
+		if !metav1.IsControlledBy(existingDNSEndpoint, vs) {
+			glog.V(3).Infof("external DNS endpoint resource isnot owned by this object. refusing to update non-owned resource")
+			return nil, nil, nil
+		}
+		if !extdnsendpointNeedsUpdate(existingDNSEndpoint, dnsEndpoint) {
```

Comment:
I would change this function name to `isUpdateRequired` since this function is already in the context of ExternalDNS